`--dart-define-from-file` entries are passed as individual properties (`-P`) to gradle

### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Create an empty project and `cd` into it

    ```
    flutter create example --platforms android --empty
    ```

1. Create a `defines.json` file in the project root:

    ```bash
    echo '{"FIRST_KEY": "FIRST_VALUE"}' > defines.json
    ```

1. Modify `app/build.gradle` to print the dart define during Gradle build. For example, add the following line inside the `android` block:

    ```gradle
    android {
        ...

        println "dart-define: FIRST_KEY = ${getProperty('FIRST_KEY')}"

        defaultConfig {
            // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
            applicationId "com.example.example"
            // You can update the following values to match your application needs.
            // For more information, see: https://docs.flutter.dev/deployment/android#reviewing-the-gradle-build-configuration.
            minSdkVersion flutter.minSdkVersion
            targetSdkVersion flutter.targetSdkVersion
            versionCode flutterVersionCode.toInteger()
            versionName flutterVersionName
        }

        ...
        ```

1. Run the build, passing the `--dart-define-from-file` option:

    ```
    flutter build apk --debug --dart-define-from-file defines.json
    ```

    It should pass - here's my output:

    ```console
    $ flutter build apk --debug --dart-define-from-file defines.json
    
    dart-define: FIRST_KEY = FIRST_VALUE
    Running Gradle task 'assembleDebug'...                           1,313ms
    ✓  Built build/app/outputs/flutter-apk/app-debug.apk.
    ```

1. Run the build, using the `--dart-define` option:

    ```
    flutter build apk --debug --dart-define 'FIRST_KEY=FIRST_VALUE'
    ```

    It fails with the following error:
    
    <details><summary>See logs</summary>
    <p>
    
    ```
    $ flutter build apk --debug --dart-define 'FIRST_KEY=First value'
    
    
    FAILURE: Build failed with an exception.
    
    * Where:
    Build file '/Users/bartek/dev/random/example/android/app/build.gradle' line: 46
    
    * What went wrong:
    A problem occurred evaluating project ':app'.
    > Could not get unknown property 'FIRST_KEY' for extension 'android' of type com.android.build.gradle.internal.dsl.BaseAppModuleExtension.
    
    * Try:
    > Run with --stacktrace option to get the stack trace.
    > Run with --info or --debug option to get more log output.
    > Run with --scan to get full insights.
    
    * Get more help at https://help.gradle.org
    
    BUILD FAILED in 436ms
    Running Gradle task 'assembleDebug'...                             937ms
    Gradle task assembleDebug failed with exit code 1
    ```
    
    </p>
    </details> 

### Expected results

I expect both ways of passing dart defines to work the same way. There's no information in the help text for `flutter build` (and friends) about this behavior:

<details><summary>flutter build apk --help</summary>
<p>

```
...
    --dart-define=<foo=bar>                             Additional key-value pairs that will be available as constants from the
                                                        String.fromEnvironment, bool.fromEnvironment, int.fromEnvironment, and
                                                        double.fromEnvironment constructors.
                                                        Multiple defines can be passed by repeating "--dart-define" multiple times.
    --dart-define-from-file=<use-define-config.json>    The path of a json format file where flutter define a global constant pool. Json
                                                        entry will be available as constants from the String.fromEnvironment,
                                                        bool.fromEnvironment, int.fromEnvironment, and double.fromEnvironment constructors;
                                                        the key and field are json values.
                                                        Multiple defines can be passed by repeating "--dart-define-from-file" multiple
                                                        times.
...
```
</p>
</details> 

### Actual results

The behavior differs.

### Code sample

N/A / provided above


### Screenshots or Video

N/A

### Logs

N/A

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.10.5, on macOS 13.4.1 22F82 darwin-arm64, locale en-PL)
    • Flutter version 3.10.5 on channel stable at /Users/bartek/fvm/versions/3.10.5
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 796c8ef792 (4 weeks ago), 2023-06-13 15:51:02 -0700
    • Engine revision 45f6e00911
    • Dart version 3.0.5
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /Users/bartek/androidsdk
    • Platform android-33, build-tools 33.0.1
    • ANDROID_HOME = /Users/bartek/androidsdk
    • Java binary at: /Users/bartek/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.10121639/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/bartek/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.10121639/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.4)
    • IntelliJ at /Users/bartek/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 74.0.4
    • Dart plugin version 231.9161.14

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.4)
    • IntelliJ at /Users/bartek/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.9225.16/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.80.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.68.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.4.1 22F82 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

--- a/packages/flutter_tools/lib/src/build_info.dart
+++ b/packages/flutter_tools/lib/src/build_info.dart
@@ -39,7 +39,6 @@ class BuildInfo {
     this.webRenderer = WebRendererMode.auto,
     required this.treeShakeIcons,
     this.performanceMeasurementFile,
-    this.dartDefineConfigJsonMap = const <String, Object?>{},
     this.packagesPath = '.dart_tool/package_config.json', // TODO(zanderso): make this required and remove the default.
     this.nullSafetyMode = NullSafetyMode.sound,
     this.codeSizeDirectory,
@@ -140,17 +139,6 @@ class BuildInfo {
   /// This is not considered a build input and will not force assemble to
   /// rerun tasks.
   final String? performanceMeasurementFile;
-
-  /// Configure a constant pool file.
-  /// Additional constant values to be made available in the Dart program.
-  ///
-  /// These values can be used with the const `fromEnvironment` constructors of
-  ///  [String] the key and field are json values
-  /// json value
-  ///
-  /// An additional field `dartDefineConfigJsonMap` is provided to represent the native JSON value of the configuration file
-  ///
-  final Map<String, Object?> dartDefineConfigJsonMap;
 
   /// If provided, an output directory where one or more v8-style heap snapshots
   /// will be written for code size profiling.
@@ -277,11 +265,7 @@ class BuildInfo {
   ///
   /// Fields that are `null` are excluded from this configuration.
   Map<String, String> toEnvironmentConfig() {
-    final Map<String, String> map = <String, String>{};
-    dartDefineConfigJsonMap.forEach((String key, Object? value) {
-      map[key] = '$value';
-    });
-    final Map<String, String> environmentMap = <String, String>{
+    return <String, String>{
       if (dartDefines.isNotEmpty)
         'DART_DEFINES': encodeDartDefines(dartDefines),
       'DART_OBFUSCATION': dartObfuscation.toString(),
@@ -303,23 +287,13 @@ class BuildInfo {
       if (codeSizeDirectory != null)
         'CODE_SIZE_DIRECTORY': codeSizeDirectory!,
     };
-    map.forEach((String key, String value) {
-      if (environmentMap.containsKey(key)) {
-        globals.printWarning(
-            'The key: [$key] already exists, you cannot use environment variables that have been used by the system!');
-      } else {
-        // System priority is greater than user priority
-        environmentMap[key] = value;
-      }
-    });
-    return environmentMap;
   }
 
   /// Convert this config to a series of project level arguments to be passed
   /// on the command line to gradle.
   List<String> toGradleConfig() {
     // PACKAGE_CONFIG not currently supported.
-    final List<String> result = <String>[
+    return <String>[
       if (dartDefines.isNotEmpty)
         '-Pdart-defines=${encodeDartDefines(dartDefines)}',
       '-Pdart-obfuscation=$dartObfuscation',
@@ -342,16 +316,6 @@ class BuildInfo {
       for (final String projectArg in androidProjectArgs)
         '-P$projectArg',
     ];
-    final Iterable<String> gradleConfKeys = result.map((final String gradleConf) => gradleConf.split('=')[0].substring(2));
-    dartDefineConfigJsonMap.forEach((String key, Object? value) {
-      if (gradleConfKeys.contains(key)) {
-        globals.printWarning(
-            'The key: [$key] already exists, you cannot use gradle variables that have been used by the system!');
-      } else {
-        result.add('-P$key=$value');
-      }
-    });
-    return result;
   }
 }
 

--- a/packages/flutter_tools/lib/src/commands/assemble.dart
+++ b/packages/flutter_tools/lib/src/commands/assemble.dart
@@ -260,7 +260,7 @@ class AssembleCommand extends FlutterCommand {
 
     final Map<String, Object?> defineConfigJsonMap = extractDartDefineConfigJsonMap();
     final List<String> dartDefines = extractDartDefines(defineConfigJsonMap: defineConfigJsonMap);
-    if (dartDefines.isNotEmpty){
+    if (dartDefines.isNotEmpty) {
       results[kDartDefines] = dartDefines.join(',');
     }
 

--- a/packages/flutter_tools/lib/src/runner/flutter_command.dart
+++ b/packages/flutter_tools/lib/src/runner/flutter_command.dart
@@ -1304,7 +1304,6 @@ abstract class FlutterCommand extends Command<void> {
       dartExperiments: experiments,
       webRenderer: webRenderer,
       performanceMeasurementFile: performanceMeasurementFile,
-      dartDefineConfigJsonMap: defineConfigJsonMap,
       packagesPath: packagesPath ?? globals.fs.path.absolute('.dart_tool', 'package_config.json'),
       nullSafetyMode: nullSafetyMode,
       codeSizeDirectory: codeSizeDirectory,

--- a/packages/flutter_tools/test/general.shard/build_info_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_info_test.dart
@@ -242,7 +242,7 @@ void main() {
     const BuildInfo buildInfo = BuildInfo(BuildMode.debug, '',
       treeShakeIcons: true,
       trackWidgetCreation: true,
-      dartDefines: <String>['foo=2', 'bar=2', 'baz=2'],
+      dartDefines: <String>['foo=2', 'bar=2'],
       dartObfuscation: true,
       splitDebugInfoPath: 'foo/',
       frontendServerStarterPath: 'foo/bar/frontend_server_starter.dart',

--- a/packages/flutter_tools/test/general.shard/build_info_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_info_test.dart
@@ -242,8 +242,7 @@ void main() {
     const BuildInfo buildInfo = BuildInfo(BuildMode.debug, '',
       treeShakeIcons: true,
       trackWidgetCreation: true,
-      dartDefines: <String>['foo=2', 'bar=2'],
-      dartDefineConfigJsonMap: <String, Object>{'baz': '2'},
+      dartDefines: <String>['foo=2', 'bar=2', 'baz=2'],
       dartObfuscation: true,
       splitDebugInfoPath: 'foo/',
       frontendServerStarterPath: 'foo/bar/frontend_server_starter.dart',
@@ -268,7 +267,6 @@ void main() {
       '-Pcode-size-directory=foo/code-size',
       '-Pfoo=bar',
       '-Pfizz=bazz',
-      '-Pbaz=2',
     ]);
   });
 
@@ -296,71 +294,5 @@ void main() {
     expect(decodeDartDefines(<String, String>{
       kDartDefines: 'MTIzMiw0NTY=,Mg==',
     }, kDartDefines), <String>['1232,456', '2']);
-  });
-
-  group('Check repeated buildInfo variables', () {
-    testUsingContext('toEnvironmentConfig repeated variable', () async {
-      const BuildInfo buildInfo = BuildInfo(BuildMode.debug, '',
-          treeShakeIcons: true,
-          trackWidgetCreation: true,
-          dartDefines: <String>['foo=2', 'bar=2'],
-          dartDefineConfigJsonMap: <String, Object>{'DART_DEFINES': 'Define a variable, but it occupies the variable name of the system'},
-          dartObfuscation: true,
-      );
-      buildInfo.toEnvironmentConfig();
-      expect(testLogger.warningText, contains('The key: [DART_DEFINES] already exists, you cannot use environment variables that have been used by the system'));
-    });
-
-    testUsingContext('toEnvironmentConfig repeated variable with DART_DEFINES not set', () async {
-      // Simulate operation flutterCommand.getBuildInfo  with `dart-define-from-file` set dartDefines
-      const BuildInfo buildInfo = BuildInfo(BuildMode.debug, '',
-          treeShakeIcons: true,
-          dartDefines: <String>['DART_DEFINES=Define a variable, but it occupies the variable name of the system'],
-          trackWidgetCreation: true,
-          dartDefineConfigJsonMap: <String, Object>{ 'DART_DEFINES' : 'Define a variable, but it occupies the variable name of the system'},
-          dartObfuscation: true,
-      );
-      buildInfo.toEnvironmentConfig();
-      expect(testLogger.warningText, contains('The key: [DART_DEFINES] already exists, you cannot use environment variables that have been used by the system'));
-
-    });
-
-    testUsingContext('toGradleConfig repeated variable', () async {
-      const BuildInfo buildInfo = BuildInfo(BuildMode.debug, '',
-          treeShakeIcons: true,
-          trackWidgetCreation: true,
-          dartDefines: <String>['foo=2', 'bar=2'],
-          dartDefineConfigJsonMap: <String, Object>{'dart-defines': 'Define a variable, but it occupies the variable name of the system'},
-          dartObfuscation: true,
-      );
-      buildInfo.toGradleConfig();
-      expect(testLogger.warningText, contains('The key: [dart-defines] already exists, you cannot use gradle variables that have been used by the system'));
-    });
-
-    testUsingContext('toGradleConfig repeated variable with not set', () async {
-      // Simulate operation flutterCommand.getBuildInfo  with `dart-define-from-file` set dartDefines
-      const BuildInfo buildInfo = BuildInfo(BuildMode.debug, '',
-          treeShakeIcons: true,
-          trackWidgetCreation: true,
-          dartDefines: <String>['dart-defines=Define a variable, but it occupies the variable name of the system'],
-          dartDefineConfigJsonMap: <String, Object>{'dart-defines': 'Define a variable, but it occupies the variable name of the system'},
-          dartObfuscation: true,
-      );
-      buildInfo.toGradleConfig();
-      expect(testLogger.warningText, contains('The key: [dart-defines] already exists, you cannot use gradle variables that have been used by the system'));
-    });
-
-    testUsingContext('toGradleConfig with androidProjectArgs override gradle project variant', () async {
-      const BuildInfo buildInfo = BuildInfo(BuildMode.debug, '',
-          treeShakeIcons: true,
-          trackWidgetCreation: true,
-          androidProjectArgs: <String>['applicationId=com.google'],
-          dartDefineConfigJsonMap: <String, Object>{'applicationId': 'override applicationId'},
-          dartObfuscation: true,
-      );
-      buildInfo.toGradleConfig();
-      expect(testLogger.warningText, contains('The key: [applicationId] already exists, you cannot use gradle variables that have been used by the system'));
-    });
-
   });
 }

